wagtail/tests/test_rich_text.py

```diff
@@ -101,6 +119,78 @@ def test_expand_db_html_with_embed(self, get_embed):
         result = expand_db_html(html)
         self.assertIn("test html", result)
 
+    # Override CACHES so we don't generate any cache-related SQL queries
+    # for page site root paths (tests use DatabaseCache otherwise).
+    @override_settings(
+        CACHES={
+            "default": {
+                "BACKEND": "django.core.cache.backends.locmem.LocMemCache",
+            },
+        }
+    )
+    def test_expand_db_html_database_queries_pages(self):
+        Site.clear_site_root_paths_cache()
+
+        with self.assertNumQueries(5):
+            expand_db_html(
+                """
+This rich text has 8 page links, and this test verifies that the code uses the
+minimal number of database queries (5) to expand them.
+
+All of these pages should be retrieved with 4 queries, one to do the base
+Page table lookup and then 1 each for the EventIndex, EventPage, and
+SimplePage tables.
+
+<a linktype="page" id="3">This links to an EventIndex page.</a>
+<a linktype="page" id="4">This links to an EventPage page.</a>
+<a linktype="page" id="5">This links to an EventPage page.</a>
+<a linktype="page" id="6">This links to an EventPage page.</a>
+<a linktype="page" id="9">This links to an EventPage page.</a>
+<a linktype="page" id="12">This links to an EventPage page.</a>
+<a linktype="page" id="7">This links to a SimplePage page.</a>
+<a linktype="page" id="11">This links to a SimplePage page.</a>
+
+Finally there's one additional query needed to do the Site root paths lookup.
+        """
+            )
+
+    def test_expand_db_html_database_queries_documents(self):
+        with self.assertNumQueries(1):
+            expand_db_html(
+                html="""
+This rich text has 2 document links, and this test verifies that the code uses
+the minimal number of database queries (1) to expand them.
+
+Both of these documents should be retrieved with 1 query:
+
+<a linktype="document" id="1">This links to a document.</a>
+<a linktype="document" id="2">This links to another document.</a>
+"""
+            )
+
+    # Disable rendition cache that might be populated by other tests.
+    @override_settings(
+        CACHES={
+            "renditions": {
+                "BACKEND": "django.core.cache.backends.dummy.DummyCache",
+            },
+        }
+    )
+    def test_expand_db_html_database_queries_images(self):
+        with self.assertNumQueries(3):
+            expand_db_html(
+                """
+This rich text has 2 image links, and this test verifies that the code uses the
+minimal number of database queries (3) to expand them.
+
+Both of these images should be retrieved with 3 queries, one to fetch the
+image objects in bulk and then one per image to fetch their renditions:
+
+This is an image: <embed embedtype="image" id="1" format="left" />
+This is another image: <embed embedtype="image" id="2" format="left" />
+        """
+            )
+
```

Comment:
The embed code doesn't currently support bulk fetching of embeds; see for example [embeds.py](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/embeds.py).

[The current embed tests](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/tests/test_rich_text.py) should properly test the existing rich text conversion (including the parts touched by this PR).

Adding a test specifically to count database queries when doing embed rich text conversion didn't seem valuable to me because that code [does a query](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/embeds.py#L17) and then [does a save](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/embeds.py#L62) (which happens in a transaction), so using `assertNumQueries` doesn't feel particularly useful.